backend/repository/handler.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package repository
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Handler provides http handlers for modifying gateway configurations.
+type Handler struct {
+	Manager       *Manager
+	DiffCreater   DiffCreater
+	gatewayHeader string
+	logger        Logger
+}
+
+// DiffCreater creates and lands a diff from changes of local repository.
+type DiffCreater interface {
+	// NewDiff return a URI for the diff or an error.
+	NewDiff(r *Repository, request *DiffRequest) (string, error)
+	// LandDiff lands a diff into a remote repository.
+	LandDiff(r *Repository, diffURI string) error
+}
+
+// Logger is based on the interface of Zap logger.
+type Logger interface {
+	Info(msg string, fields ...zapcore.Field)
+	Error(msg string, fields ...zapcore.Field)
+}
+
+// DiffRequest contains all information to create a diff.
+type DiffRequest struct {
+	BranchName    string
+	CommitMessage string
+	Reviewers     []string
+}
+
+// NewHandler constructs a new Handler.
+func NewHandler(m *Manager, dc DiffCreater, gatewayHeader string, logger Logger) *Handler {
+	return &Handler{
+		Manager:       m,
+		DiffCreater:   dc,
+		gatewayHeader: gatewayHeader,
+		logger:        logger,
+	}
+}
+
+// NewHTTPRouter constructs the endpoints for http server.
+func (h *Handler) NewHTTPRouter() *httprouter.Router {
```

Comment:
is it possible to create a `HandlerServer` instead of a `Handler`

We create a function `NewRepositoryHandler` server that takes some options ( like a port / ip ) and starts a gateway and initializes all of the things internally.

zanzibar has evolved away from gateway to service baseline, we should look into using it for edge-gateway-orchestrator.

I think it's out of scope for this PR as we might need extra refactoring in zanzibar/runtime/ to make it compatible with edge gateway orchestrator